pkg/patroni/adapter.go

```diff
@@ -0,0 +1,711 @@
+	}
+
+	// Build the JSON payload in Patroni's expected format
+	patchRequest := PatroniPatchRequest{}
+	patchRequest.PostgreSQL.Parameters = pgParameters
+
+	// Convert to JSON
+	jsonData, err := json.Marshal(patchRequest)
+	if err != nil {
+		logger.Errorf("Error marshaling JSON: %v", err)
+		return fmt.Errorf("failed to marshal JSON: %w", err)
+	}
+
+	logger.Debugf("Patroni PATCH payload: %s", string(jsonData))
+
+	// Step 3: Apply configuration via Patroni REST API
+	configURL := fmt.Sprintf("%s/config", adapter.PatroniConfig.PatroniAPIURL)
+
+	// Create HTTP PATCH request
+	req, err := http.NewRequest("PATCH", configURL, bytes.NewBuffer(jsonData))
+	if err != nil {
+		logger.Errorf("Error creating request: %v", err)
+		return fmt.Errorf("failed to create HTTP request: %w", err)
+	}
+
+	// Set Content-Type header
+	req.Header.Set("Content-Type", "application/json")
+
+	// Send the request
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		logger.Errorf("Error sending request: %v", err)
+		return fmt.Errorf("failed to send HTTP PATCH request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// Read response body
+	responseBody, err := io.ReadAll(resp.Body)
+	if err != nil {
+		logger.Errorf("Error reading response: %v", err)
+		return fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	logger.Infof("Patroni API Response Status: %s", resp.Status)
+	logger.Debugf("Patroni API Response Body: %s", string(responseBody))
+
+	// Check if the request was successful (2xx status codes)
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		logger.Errorf("Patroni API returned error status: %d, body: %s", resp.StatusCode, string(responseBody))
+		return fmt.Errorf("patroni API returned error status %d: %s", resp.StatusCode, string(responseBody))
+	}
+
+	logger.Info("Configuration successfully applied via Patroni REST API")
+
+	// Step 4: Wait for PostgreSQL if restart is needed
+	if proposedConfig.KnobApplication == "restart" {
+		logger.Warn("Configuration requires restart, waiting for PostgreSQL to be ready...")
+		err = pg.WaitPostgresReady(adapter.PGDriver)
+		if err != nil {
+			return fmt.Errorf("failed to wait for PostgreSQL to be back online: %w", err)
+		}
+		logger.Info("PostgreSQL is ready after restart")
+	}
+
+	// Step 5: Verify that all parameters were applied correctly by reading pg_settings
+	logger.Info("Waiting for Patroni to apply configuration (15s delay)...")
+	time.Sleep(15 * time.Second)
+	logger.Info("Verifying applied configuration in pg_settings...")
+	for _, knob := range parsedKnobs {
+		// Skip verification for Patroni-managed parameters
+		if IsPatroniManagedParameter(knob.Name) {
+			logger.Warnf("Skipping verification for Patroni-managed parameter: %s", knob.Name)
+			continue
+		}
+
+		query := `SELECT setting, unit FROM pg_settings WHERE name = $1`
+		var actualSetting string
+		var unit *string
+		err := adapter.PGDriver.QueryRow(ctx, query, knob.Name).Scan(&actualSetting, &unit)
```

Comment:
Please use the `execWithPrefix` in `utils`. We need to make sure we always prefix all SQL queries with `/*dbtune*/` in a uniform way, such that it doesn't show up as queries that are tuned.